geps/gep-2649/index.md

```diff
@@ -0,0 +1,714 @@
+
+### Conflict Resolution
+
+Other conflict resolution between Policy objects must also follow the rules
+outlined in GEP-713, in its Conflict Resolution section.
+
+## Policy Attachment examples and behavior
+
+This approach is building on concepts from all of the alternatives discussed
+below. This is very similar to the (now removed) BackendPolicy resource in the API,
+but also borrows some concepts from the [ServicePolicy
+proposal](https://github.com/kubernetes-sigs/gateway-api/issues/611).
+
+### Policy Attachment for Ingress
+Attaching a Directly Attached Policy to Gateway resources for ingress use cases
+is relatively straightforward. A policy can reference the resource it wants to
+apply to.
+
+Access is granted with RBAC - anyone that has access to create a RetryPolicy in
+a given namespace can attach it to any resource within that namespace.
+
+![Simple Ingress Example](images/2649-ingress-simple.png)
+
+An Inherited Policy can attach to a parent resource, and then each policy
+applies to the referenced resource and everything below it in terms of hierarchy.
+Although this example is likely more complex than many real world
+use cases, it helps demonstrate how policy attachment can work across
+namespaces.
+
+![Complex Ingress Example](images/2649-ingress-complex.png)
+
+### Policy Attachment for Mesh
+Although there is a great deal of overlap between ingress and mesh use cases,
+mesh enables more complex policy attachment scenarios. For example, you may want
+to apply policy to requests from a specific namespace to a backend in another
+namespace.
+
+![Simple Mesh Example](images/2649-mesh-simple.png)
+
+Policy attachment can be quite simple with mesh. Policy can be applied to any
+resource in any namespace but it can only apply to requests from the same
+namespace if the target is in a different namespace.
+
+At the other extreme, policy can be used to apply to requests from a specific
+workload to a backend in another namespace. A route can be used to intercept
+these requests and split them between different backends (foo-a and foo-b in
+this case).
+
+![Complex Mesh Example](images/2649-mesh-complex.png)
+
+### Policy TargetRef API
+
+Each Policy resource MUST include a single `targetRef` field. It must not
+target more than one resource at a time, but it can be used to target larger
+resources such as Gateways or Namespaces that may apply to multiple child
+resources.
+
+As with most APIs, there are countless ways we could choose to expand this in
+the future. This includes supporting multiple targetRefs and/or label selectors.
+Although this would enable compelling functionality, it would increase the
+complexity of an already complex API and potentially result in more conflicts
+between policies. Although we may choose to expand the targeting capabilities
+in the future, at this point it is strongly preferred to start with a simpler
+pattern that still leaves room for future expansion.
```

Comment:
This is an interesting decision, with which fundamentally I do not disagree of course. My 2 cents…

I can’t help thinking though on the implications of the `targetRef` structure not counting with the `sectionName` field as an option, unlike proposed for [Direct Policies](https://github.com/kubernetes-sigs/gateway-api/blob/a01cdd6cf0bb52dc2a9d967b11006b81a12b1239/geps/gep-2648/index.md?plain=1#L55-L58).

I understand the rationale behind this has to do with the cardinality of combinatory cases involving partial targeting of a resource on top of all other variables that come with Inherited Policies, and not because anybody would claim such use case is not needed. Otherwise it wouldn’t have been proposed for the other class of policies.

I believe that being able to target a section of a resource is as much needed for Inherited Policies as it is for Direct Policies, especially given complex objects such as Gateways (multiple listeners) and HTTPRoutes (multiple rules.) By not having such option in the `targetRef`, policy designers will sometimes have to find a way to specify that the fields declared in a metaresource have effect only at given point within the targeted object. I guess that will vary from implementation to implementation. Maybe it will be by means of merging values into those lists of complex structs using the _convention_ of the `name` field being the key of the structure, rather than as an identifier of a section _by design_.

I guess another option would be splitting the target resource into many. Of course, this approach would not pass the test where an implementation provides more than one kind of policy, after all one policy cannot target multiple objects. As an example, imagine xRoute _R_ is targeted by TimeoutPolicy _T_ and now one of its rules needs to be targeted by a new AuthPolicy _A_. If you split _R_ into _R1_ and _R2_, then you need 2 copies of TimeoutPolicy _T_ as well.

One could even argue moreover that the fact itself that an object would _have to split_ into many is a design flaw. Taken to the limit, you can end up into a situation of one route object per route rule, one gateway per gateway listener, etc.

Not having the `sectionName` in the `targetRef` is also inconsistent with the design for Direct Policies, which makes evolving policies from one class into the other not straightforward.

Again, I obviously won’t deny the cognitive complexity involved in computing and/or reasoning on the final effective policy while cascading throughout the hierarchy, with sets and subsets to work out along the way, and also that starting small and slowly expanding functionalities is a legit strategy. In this particular case however, for the sake of consistency between the APIs or simply because it’s needed in some cases and quite easily the lack of direction could lead to implementation chaos, I think that having the `sectionName` in the `targetRef` should be a choice of one who designs a metaresource and not a restriction imposed by the framework.

Reply:
Your points are fair, but I am _super_ worried about the complexity problem with `sectionName` for Inherited Policies.

Here I'm trying to go with "no is temporary, yes is forever", and not do it _yet_. If someone can come up with a way to manage the complexity that will result from having multiple Inherited policies, possibly with defaults and overrides, that also only target _part_ of a resource, and _still_ make it so that the owners of the affected resources can understand the system, without breaking the apiserver with constant status update load, then I'd love to say yes.

But that is a _lot_ of things that need to line up, so I'd prefer to leave this as is _for now_.

Reply:
Fair enough, @youngnick. I think your concerns are well-justified. At Kuadrant, we'll continue to try to find a solution for the problem.

API-wise, having the option to use the `sectionName` in the `targetRef` would open up this path to explore (without breaking apart from GEP-2649.) We were hoping to obtain simpler policies and simpler status blocks as well.

Our other option is to use the policy spec – I mean the part that is not `targetRef` obviously. Maybe users will write bigger policies, with bits that sometimes apply to the entire target, sometimes only to parts of it. The status block of the policy will probably be a bit messier.

Either way, targeting only a part of an object is sometimes unavoidable, I believe. Especially when the domain is auth (different authz rules for different route endpoints – e.g. admin, non-admin, anonymous access, etc) and rate-limit (again, different rate limits for different API endpoints – e.g. more expensive operation, less expensive operation), Kuadrant's kinds of Inherited Policies.

Reply:
Look, I can't stop you from using the `targetRef` variant that has `sectionName`. It's your policy. But I suppose I could make this "Inherited Policies SHOULD NOT use `sectionName` " instead if you really want to be covered.

But you could go and do this right now, and then we can change the document later. This is still Experimental, so that kind of change is okay.

Reply:
I reread this now, and I think I explained myself poorly.

What I'm saying is; it's _fine_ for you to go and build an Inherited Policy that uses `sectionName` to try things out. It will _currently_ be not compliant to the spec, but in the event you find that the discoverability and complexity problems are not as bad as we fear, then we can update this with your experience.

I would, however, prefer not to block this merging on this specific thing. Like I said earlier, this is "not supported _for now_". But I'm happy to work with you on what doing `sectionName` on Inherited Policies could look like.

Reply:
I honestly think that the change introduced in https://github.com/kubernetes-sigs/gateway-api/pull/2442 to allow some form of inheritance precedence logic on **Direct** policies is misplaced, and should instead be allowed **only** for **Inherited** policies (or a single consolidated policy type), while it could be reasonable for Direct policies to raise a `Conflicted` status if multiple instances of the same policy type were to target overlapping sections of the same resource.

Targeting separate _sibling_ sections of the same resource with Direct policies (e.g. different policies of the same type each targeting a different Listener on a Gateway) feels reasonable, but introducing a limited form of inheritance on direct policies where targeting both a Gateway and a specific Listener on it _without_ being able to tweak advanced details like merge strategy feels like a mistake.

IMO `sectionName` actually fits relatively neatly and clearly into the existing hierarchy as:
- GatewayClass
- Gateway Namespace
- Gateway
- Gateway Listener (`sectionName`)
- Route Namespace
- Route
- Route Rule (`sectionName`, with #995 merged)

Reply:
Opened https://github.com/kubernetes-sigs/gateway-api/discussions/2927#discussioncomment-9024156 to continue this discussion.